{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "from bm_tools.utils import Utils\n",
    "\n",
    "import os\n",
    "import openpyxl\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from bm_tools.data import Data\n",
    "from bm_tools.stats import Stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = \"./data/datasets\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data shapes: sample_data : (18, 7285), sample_metadata : (18, 1), feature_metadata : (7285, 3)\n"
     ]
    }
   ],
   "source": [
    "# load proteomics data and reshape to data matrix\n",
    "data_path = \"./data_Proteome/proteomics_filtered.txt\"\n",
    "df = pd.read_csv(data_path, sep=\"\\t\")\n",
    "\n",
    "# index\n",
    "sample_data = df.copy()\n",
    "sample_data.set_index(\"ProteinGroups\", inplace = True, drop = False)\n",
    "\n",
    "# sample data\n",
    "sample_data = sample_data.drop(['ProteinGroups', 'Genes', 'ProteinDescriptions'], axis = 1).T\n",
    "sample_data.index = sample_data.index.str.replace('ifng', 'ifng_').str.replace('control', 'control_')\n",
    "\n",
    "# feature metadata\n",
    "feature_metadata = df[['ProteinGroups', 'Genes', 'ProteinDescriptions']].copy()\n",
    "feature_metadata.set_index(\"ProteinGroups\", inplace = True, drop = False)\n",
    "\n",
    "# sample metadata\n",
    "sample_metadata = pd.DataFrame({\"sample\" : sample_data.index})\n",
    "sample_metadata.set_index(\"sample\", inplace = True, drop = False)\n",
    "\n",
    "print(f\"Data shapes: sample_data : {sample_data.shape}, sample_metadata : {sample_metadata.shape}, feature_metadata : {feature_metadata.shape}\")\n",
    "\n",
    "# create Data instance to check alignment, apply filtering, track metadata etc. \n",
    "data = Data()\n",
    "data.add_anndata(sample_data)\n",
    "data.add_metadata(sample_metadata, axis = 0)\n",
    "data.add_metadata(feature_metadata, axis = 1)\n",
    "\n",
    "# save data\n",
    "data.save_dataframes(\n",
    "    directory = data_dir,\n",
    "    name = \"islets_proteomics\",\n",
    "    pickle = True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(17230, 35)\n",
      "Shape before deduplication: (13098, 21), shape after deduplication: (12993, 21), aggregated 105 rows.\n",
      "Data shapes: sample_data : (18, 12993), sample_metadata : (18, 1), feature_metadata : (12993, 3)\n"
     ]
    }
   ],
   "source": [
    "# load transcriptomics data and reshape to data matrix\n",
    "# data_path = \"./data_RNAseq/MARINE_ilots_ifng_expression_DeSeq2.xlsx\" # old dataset\n",
    "data_path = \"./data_RNAseq/ilots_ifng_expression_DeSeq2_Marine_20220610.xlsx\"\n",
    "df = pd.read_excel(data_path, engine = 'openpyxl')\n",
    "print(df.shape)\n",
    "#TODO: figure out strange openpyxl/pandas warning for old dataset\n",
    "\n",
    "# remove unused columns\n",
    "mean_ratio_columns = [\n",
    "    'alpha.ctrl',\n",
    "    'alpha.IFNg',\n",
    "    'alpha.log2fc',\n",
    "    'alpha.p.value', \n",
    "    'alpha.p.adj',\n",
    "    'beta.ctrl',\n",
    "    'beta.IFNg',\n",
    "    'beta.log2fc',\n",
    "    'beta.p.value', \n",
    "    'beta.p.adj',\n",
    "    'delta.ctrl',\n",
    "    'delta.IFNg',\n",
    "    'delta.log2fc', \n",
    "    'delta.p.value', \n",
    "    'delta.p.adj'\n",
    "    ]\n",
    "df.drop(mean_ratio_columns, axis = 1, inplace = True)\n",
    "\n",
    "# # add uniprot id column to ENSEMBL id column\n",
    "# mmus_ref_dict = Utils.map_ensembl_to_uniprot('mmusculus', True)\n",
    "\n",
    "# # Save the mmus_ref_dict to a pickle file\n",
    "# NOW = pd.Timestamp.now().strftime(\"%Y-%m-%d_%H_%M_%S\")\n",
    "# mmus_ref_dict_df = pd.DataFrame.from_dict(mmus_ref_dict, orient = 'index', columns = ['uniprot_id'])\n",
    "# mmus_ref_dict_df.to_pickle(f\"./{NOW}_mmus_ref_dict.pkl\")\n",
    "\n",
    "# load mmus_ref_dict from pickle file\n",
    "mmus_ref_dict = pd.read_pickle(\"./2025-04-23_12_01_01_mmus_ref_dict.pkl\")['uniprot_id'].to_dict()\n",
    "\n",
    "df['uniprot_id'] = df['ensembl.id'].map(mmus_ref_dict)\n",
    "df.set_index('uniprot_id', inplace = True, drop = False)\n",
    "\n",
    "# run checks: remove rows with missing uniprot ids\n",
    "df.dropna(subset = ['uniprot_id'], inplace = True)\n",
    "\n",
    "# deduplicate by aggregation: either average or sum per uniprot id. String columns are concatenated with a semicolon\n",
    "df = Utils.deduplicate_alphanumeric_dataframe(df, 'sum')\n",
    "\n",
    "# index\n",
    "sample_data = df.copy()\n",
    "sample_data.set_index('uniprot_id', inplace = True, drop = False)\n",
    "\n",
    "# sample data\n",
    "sample_data = sample_data.drop(['ensembl.id', 'gene.symbol', 'uniprot_id'], axis = 1).T\n",
    "sample_data.index = sample_data.index.str.lower()\n",
    "\n",
    "# sample metadata\n",
    "sample_metadata = pd.DataFrame({\"sample\" : sample_data.index})\n",
    "sample_metadata.set_index(\"sample\", inplace = True, drop = False)\n",
    "\n",
    "# feature metadata\n",
    "feature_metadata = df[['ensembl.id', 'gene.symbol', 'uniprot_id']].copy()\n",
    "feature_metadata.set_index('uniprot_id', inplace = True, drop = False)\n",
    "\n",
    "print(f\"Data shapes: sample_data : {sample_data.shape}, sample_metadata : {sample_metadata.shape}, feature_metadata : {feature_metadata.shape}\")\n",
    "\n",
    "# create Data instance to check alignment, apply filtering, track metadata etc.\n",
    "data = Data()\n",
    "data.add_anndata(sample_data)\n",
    "data.add_metadata(sample_metadata, axis = 0)\n",
    "data.add_metadata(feature_metadata, axis = 1)\n",
    "\n",
    "# apply log2 transformation to transcriptomics data\n",
    "data.log_transform(log = 2)\n",
    "\n",
    "# save data\n",
    "data.save_dataframes(\n",
    "    directory = data_dir,\n",
    "    name = \"islets_transcriptomics\",\n",
    "    pickle = True\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data shapes: sample_data : (9, 7356), sample_metadata : (9, 7), feature_metadata : (7356, 25)\n"
     ]
    }
   ],
   "source": [
    "# load phospho data and reshape to data matrix\n",
    "data_path = \"./data_Phospho/celltype_psite.txt\"\n",
    "df = pd.read_csv(data_path, sep=\"\\t\")\n",
    "\n",
    "# index\n",
    "sample_data = df.copy()\n",
    "sample_data.set_index('PTM_collapse_key', inplace = True, drop = False)\n",
    "\n",
    "# sample data\n",
    "sample_data = sample_data[[\n",
    "    '20230426_TIMS01scp_MCT_SA_whi20_AID12_DBS08_Phospho_07_alpha_01_S3-G1_1_2886',\n",
    "    '20230426_TIMS01scp_MCT_SA_whi20_AID12_DBS08_Phospho_08_alpha_02_S3-H1_1_2887',\n",
    "    '20230426_TIMS01scp_MCT_SA_whi20_AID12_DBS08_Phospho_09_alpha_03_S3-A2_1_2888',\n",
    "    '20230426_TIMS01scp_MCT_SA_whi20_AID12_DBS08_Phospho_10_beta_01_S3-B2_1_2889',\n",
    "    '20230426_TIMS01scp_MCT_SA_whi20_AID12_DBS08_Phospho_11_beta_02_S3-C2_1_2890',\n",
    "    '20230426_TIMS01scp_MCT_SA_whi20_AID12_DBS08_Phospho_12_beta_03_S3-D2_1_2891',\n",
    "    '20230426_TIMS01scp_MCT_SA_whi20_AID12_DBS08_Phospho_13_delta_01_S3-E2_1_2883',\n",
    "    '20230426_TIMS01scp_MCT_SA_whi20_AID12_DBS08_Phospho_14_delta_02_S3-F2_1_2884',\n",
    "    '20230426_TIMS01scp_MCT_SA_whi20_AID12_DBS08_Phospho_15_delta_03_S3-G2_1_2885',]].T\n",
    "\n",
    "# sample_metadata\n",
    "sample_metadata = pd.DataFrame({\"sample\" : sample_data.index})\n",
    "sample_metadata.set_index(\"sample\", inplace = True, drop = False)\n",
    "sample_metadata['treat'] = 'control'\n",
    "sample_metadata['sample'] = sample_metadata.index.str.split('_').str[-5]\n",
    "sample_metadata['rep'] = sample_metadata.index.str.split('_').str[-4].astype(int).astype(str)\n",
    "sample_metadata['readout'] = 'phosphoproteomics'\n",
    "sample_metadata['sample_treat_readout'] = sample_metadata['sample'] + '_' + sample_metadata['treat'] + '_' + sample_metadata['readout']\n",
    "sample_metadata['sample_treat_rep_readout'] = sample_metadata['sample'] + '_' + sample_metadata['treat'] + '_' + sample_metadata['rep'] + '_' + sample_metadata['readout']\n",
    "\n",
    "# replace data index with matched metadata sample_treat_rep_readout\n",
    "sample_data = sample_data.join(sample_metadata['sample_treat_rep_readout'])\n",
    "sample_data.set_index('sample_treat_rep_readout', inplace = True, drop = True)\n",
    "\n",
    "# replace sample metadata index with sample_treat_rep_readout\n",
    "sample_metadata['PTM_collapse_key'] = sample_metadata.index\n",
    "sample_metadata.set_index('sample_treat_rep_readout', inplace = True, drop = False)\n",
    "\n",
    "# feature metadata\n",
    "feature_metadata = df[['R.Condition', 'PG.Genes', 'PG.Organisms', 'PG.ProteinDescriptions',\n",
    "       'PG.ProteinGroups', 'PG.ProteinNames', 'PG.UniProtIds',\n",
    "       'PEP.PeptidePosition', 'EG.IsDecoy', 'EG.PrecursorId',\n",
    "       'EG.PTMPositions..Phospho..STY..', 'EG.ApexRT',\n",
    "       'EG.PTMProbabilities..Phospho..STY..', 'EG.PTMSites..Phospho..STY..',\n",
    "       'EG.PTMAssayCandidateScore', 'EG.PTMAssayProbability',\n",
    "       'EG.PTMLocalizationProbabilities', 'EG.ProteinPTMLocations',\n",
    "       'EG.NormalizationFactor', 'PTM_0_num', 'PTM_group', 'PTM_collapse_key',\n",
    "       'PTM_collapse_key_num', 'PTM_localization', 'PTM_0_aa']].copy()\n",
    "feature_metadata.set_index('PTM_collapse_key', inplace = True, drop = False)\n",
    "\n",
    "print(f\"Data shapes: sample_data : {sample_data.shape}, sample_metadata : {sample_metadata.shape}, feature_metadata : {feature_metadata.shape}\")\n",
    "\n",
    "# create Data instance to check alignment, apply filtering, track metadata etc.\n",
    "data = Data()\n",
    "data.add_anndata(sample_data)\n",
    "data.add_metadata(sample_metadata, axis = 0)\n",
    "data.add_metadata(feature_metadata, axis = 1)\n",
    "\n",
    "# save data\n",
    "data.save_dataframes(\n",
    "    directory = data_dir,\n",
    "    name = \"islets_phospho\",\n",
    "    pickle = True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save usable datasets for proteomics and transcriptomics data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(36, 4481)\n"
     ]
    }
   ],
   "source": [
    "data_dir = \"./data/datasets\"\n",
    "subframes = ['alpha_control', 'beta_control', 'delta_control', 'alpha_ifng', 'beta_ifng', 'delta_ifng']\n",
    "\n",
    "# load proteomics data\n",
    "pdata = pd.read_pickle(os.path.join(data_dir, 'islets_proteomics_data.pkl'))\n",
    "\n",
    "# load transcriptomics data\n",
    "tdata = pd.read_pickle(os.path.join(data_dir, \"islets_transcriptomics_data.pkl\"))\n",
    "\n",
    "# experiment with aggregating the data and transporting it around with a Data object\n",
    "pdata.index = [x + \"_proteomics\" for x in pdata.index]\n",
    "tdata.index = [x + \"_transcriptomics\" for x in tdata.index]\n",
    "\n",
    "# combine the data\n",
    "data = pdata.T.join(tdata.T, how = 'inner').T\n",
    "print(data.shape)\n",
    "\n",
    "# split idx\n",
    "sample_treat_rep_readout = data.index\n",
    "sample = sample_treat_rep_readout.str.split(\"_\").str[0]\n",
    "treat = sample_treat_rep_readout.str.split(\"_\").str[1]\n",
    "rep = sample_treat_rep_readout.str.split(\"_\").str[2]\n",
    "readout = sample_treat_rep_readout.str.split(\"_\").str[3]\n",
    "\n",
    "# make metadata\n",
    "md = pd.DataFrame({\n",
    "    \"sample\": sample,\n",
    "    \"treat\" : treat,\n",
    "    \"rep\": rep,\n",
    "    \"readout\": readout,\n",
    "    \"sample_treat_readout\": [\"_\".join(x) for x in zip(sample, treat, readout)],\n",
    "}, index = data.index)\n",
    "\n",
    "# make a Data object\n",
    "data_obj = Data()\n",
    "data_obj.add_anndata(data)\n",
    "data_obj.add_metadata(md, 0)\n",
    "\n",
    "# save to disk\n",
    "data_obj.get_data().to_pickle(\"./rna_prot_dataset.pkl\")\n",
    "data_obj.get_metadata(0).to_pickle(\"./rna_prot_sample_metadata.pkl\")\n",
    "data_obj.get_metadata(1).to_pickle(\"./rna_prot_feature_metadata.pkl\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "momi",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
